### PR TITLE
GH-37851: [C++] IPC: ArrayLoader style enhancement

### LINKED ISSUE #37851: [C++] Would `arrow::ipc::ArrayLoader` get dangling pointer when `SkipField` used?
### Describe the enhancement requested

In `arrow::ipc::ArrayLoader`, `SkipField` will create a dummy `ArrayData`, and point `arrow::ipc::ArrayLoader::out_` into that. Would this causing dangling pointer? Or it's by design?

```c++
  Status Load(const Field* field, ArrayData* out) {
    if (max_recursion_depth_ <= 0) {
      return Status::Invalid("Max recursion depth reached");
    }

    field_ = field;
    out_ = out;
    out_->type = field_->type();
    return LoadType(*field_->type());
  }

  Status SkipField(const Field* field) {
    ArrayData dummy;
    skip_io_ = true;
    Status status = Load(field, &dummy);
    skip_io_ = false;
    return status;
  }
```

### Component(s)

C++

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

Enhance the style of `arrow::ipc::ArrayLoader`'s `SkipField`

### What changes are included in this PR?

Set `out_` to `nullptr` when `SkipField` is called.

### Are these changes tested?

No

### Are there any user-facing changes?

No

* Closes: #37851

### CODE REVIEW

**Comment by mapleFU** [comment]
@kou I've set `out_ = nullptr` and add `DCHECK` here

File: cpp/src/arrow/ipc/reader.cc [hunk 1]
```diff
@@ -208,6 +208,7 @@ class ArrayLoader {
   Status LoadType(const DataType& type) { return VisitTypeInline(type, this); }
 
   Status Load(const Field* field, ArrayData* out) {
+    DCHECK_NE(nullptr, out);
```
Comment (kou) [inline cpp/src/arrow/ipc/reader.cc:211]: Ah, I assumed that we want to add `DCHECK` to where `out_` is used something like:

```diff
diff --git a/cpp/src/arrow/ipc/reader.cc b/cpp/src/arrow/ipc/reader.cc
index 0def0e036..8b99c8321 100644
--- a/cpp/src/arrow/ipc/reader.cc
+++ b/cpp/src/arrow/ipc/reader.cc
@@ -276,6 +276,7 @@ class ArrayLoader {
 
   template <typename TYPE>
   Status LoadPrimitive(Type::type type_id) {
+    DCHECK_NE(out_, nullptr);
     out_->buffers.resize(2);
 
     RETURN_NOT_OK(LoadCommon(type_id));
```

Do we need to add these `DCHECK`s?

Comment (mapleFU) [reply cpp/src/arrow/ipc/reader.cc:211]: Oh let me add them

File: cpp/src/arrow/ipc/reader.cc [hunk 2]
```diff
@@ -290,6 +296,7 @@ class ArrayLoader {
 
   template <typename TYPE>
   Status LoadBinary(Type::type type_id) {
+    DCHECK_NE(nullptr, out_);
```
Comment (bkietz) [inline cpp/src/arrow/ipc/reader.cc:299]: Minor nit: would you putting the value checked on the left hand side?
```suggestion
    DCHECK_NE(out_, nullptr);
```

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/ipc/reader.cc [hunk 3]
```diff
@@ -243,6 +247,7 @@ class ArrayLoader {
   }
 
   Status GetFieldMetadata(int field_index, ArrayData* out) {
+    DCHECK_NE(nullptr, out_);
```
Comment (kou) [inline cpp/src/arrow/ipc/reader.cc:250]: We don't use `out_` in this method.

File: cpp/src/arrow/ipc/reader.cc (hunk 2 above)
Comment (mapleFU) [reply cpp/src/arrow/ipc/reader.cc:299]: done!

File: cpp/src/arrow/ipc/reader.cc (hunk 3 above)
Comment (mapleFU) [reply cpp/src/arrow/ipc/reader.cc]: oops, I've change the code, and push check in:
1. Before `Visit` the subtypes
2. `LoadXXX` called.

**Comment by mapleFU** [comment]
Updated @kou @bkietz , would you mind take a look again?

File: cpp/src/arrow/ipc/reader.cc (hunk 2 above)
Comment (mapleFU) [reply cpp/src/arrow/ipc/reader.cc]: Sorry, I've update it now

**Review by bkietz — APPROVED** [review]

**Pull request merged**